### PR TITLE
Add ZStream#runManaged

### LINKED ISSUE #3758: Add ZStream#runManaged
`ZStream#run[B]` produces a value and immediately closes allocated resources.
Sometimes a user wants to do something with `B` before finalizing the resources.
So `ZStream` has a bunch of `*managed` methods e.g. `foreachChunkWhileManaged` to lift limitations of `run`.
What we need:
1) add `ZStream#runManaged(sink: ZSink[B])` which returns `ZManaged[R, E, B]`
2) `ZStream.run` will just delegate to `runManaged`
3) all `ZStream#*Managed` methods  go away

### PR DESCRIPTION
**Pull request by simpadjo**
Closes #3758

After adding `ZStream#runManaged` all other `ZStream#.....Managed` methods become redundant. 
~~I left the following methods~~
~~1) that spark joy~~
~~2) that are used internally~~
~~3) that have tests~~

I kept all existing methods, made them delegating to appropriate sink

### CODE REVIEW

**Comment by simpadjo** [comment]
@iravid dotty error is irrelevant

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 1]
```diff
@@ -2292,7 +2294,10 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
    * Runs the sink on the stream to produce either the sink's result or an error.
    */
   def run[R1 <: R, E1 >: E, B](sink: ZSink[R1, E1, O, Any, B]): ZIO[R1, E1, B] =
-    (process <*> sink.push).use {
+    runManaged(sink).useNow
+
+  def runManaged[R1 <: R, E1 >: E, B](sink: ZSink[R1, E1, O, Any, B]): ZManaged[R1, E1, B] =
+    (process <*> sink.push).flatMap {
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala:2300]:
```suggestion
    (process <*> sink.push).mapM {
```

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 2]
```diff
@@ -2308,7 +2313,7 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
               .foldCauseM(c => Cause.sequenceCauseEither(c.map(_._1)).fold(IO.halt(_), ZIO.succeedNow), _ => go)
         )
 
-        go
+        go.toManaged_
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala:2316]:
```suggestion
        go
```

Comment (simpadjo) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: :+1:

**Pull request merged**